Plugins/upenn.cbica.captk.example3/src/internal/QmitkCaPTkExample3View.cpp

```
@@ -0,0 +1,205 @@
+{
+	//get datastorage( we use it further down )
+	auto ds = this->GetDataStorage();
+
+	//get selected nodes
+	QList<mitk::DataNode::Pointer> nodes = this->GetDataManagerSelection();
+
+	//we don't handle the case where data is not loaded or more than 1 nodes are selected
+	if (nodes.empty() || nodes.size() > 1)
+	{
+		QMessageBox msgError;
+		msgError.setText("Please load and select a dataset.");
+		msgError.setIcon(QMessageBox::Critical);
+		msgError.setWindowTitle("selection error");
+		msgError.exec();
+	}
+	else
+	{
+		//get first node from list
+		mitk::DataNode *node = nodes.front();
+
+		//clone the selected node
+		mitk::DataNode::Pointer clonedNode = node->Clone();
+
+		//set some new property to cloned node to distinguish it from existing node
+		clonedNode->SetProperty("name", mitk::StringProperty::New("child"));
+
+		//get data from selected node
+		mitk::BaseData* data = node->GetData();
+
+		//set it to cloned node
+		clonedNode->SetData(data);
+
+		//add cloned node to ds
+		ds->Add(clonedNode, node);
+	}
+}
+
+
+void QmitkCaPTkExample3View::OnInvertImageButtonClicked()
+{
+    //get datastorage( we use it further down )
+	auto ds = this->GetDataStorage();
+
+	//get selected nodes
+	QList<mitk::DataNode::Pointer> nodes = this->GetDataManagerSelection();
+
+	//we don't handle the case where data is not loaded or more than 1 nodes are selected
+	if (nodes.empty() || nodes.size() > 1)
+	{
+		QMessageBox msgError;
+		msgError.setText("Please load and select a dataset.");
+		msgError.setIcon(QMessageBox::Critical);
+		msgError.setWindowTitle("selection error");
+		msgError.exec();
+	}
+	else
+	{
+	    //get first node from list
+		mitk::DataNode *node = nodes.front();
+
+		auto data = node->GetData();
+		// node has data?
+		if (data != nullptr) 
+		{
+			// get smart pointer from data
+			mitk::Image::Pointer image = dynamic_cast<mitk::Image*>(data);
+			// ... has IMAGE data? :D
+			if (image.IsNotNull())
```

Maybe:

if (image.IsNotNull() && image.GetDimension() == 3)

Because it would crash with 2D. Or maybe a quick 3D check above.

---

The idea is to let it crash deliberately so you can then introduce more advanced stuff - predicates, accessbyitk etc.